Make string auto-conversion optional & use leaner strings classes

mara004 opened the pull request:
Successor PR of #162, see https://github.com/ctypesgen/ctypesgen/pull/162#issuecomment-1611149894.
This change makes string support classes optional, and replaces the current ones with something lean.
No changes beyond strings this time.

mara004 committed:
Draft: making string autoconversion optional

This is only a draft, still need to re-implement the String and
ReturnString classes.

mara004 committed:
Experiment: new replacement classes

the test suite passes, but really not sure if that is
correct/applicable, though

mara004 committed:
Different return strings approach

Previously, this would return String with underlying bytes.
Due to the wrapper, both str() and .decode() would work.

However, with plain bytes, str() does repr(), and only decodes if using
str(..., encoding="utf-8").

Not sure how to handle this nicely without API breakage.
This solution breaks str(). The previous one would break .decode().

mara004 commented:
I'm not sure how to replace the current return strings implementation without causing API breakage, see cddab30dc3c49c1451a9fa448b16136e5f348b15.
Also, if we separate input and return strings, I think we have a problem with libraries using a custom typedef for both, so a replacement would need to handle both in the same class like the current String.

For a "zero api breakage" solution, we could move the strings classes to a separate file and have them included in the preamble selectively if `use_autostrings == True`.

mara004 committed:
Combine input/return strings in one class & wrap bytes

mara004 commented:
Update to https://github.com/ctypesgen/ctypesgen/pull/177#issuecomment-1611401189:
Found out we can handle both input/return in the same class. Added a hack subclassing from bytes to support both str() and .decode(), hopefully making it effectively backwards compatible.

mara004 committed:
style nit

mara004 committed:
Apply black/flake8

mara004 commented on ctypesgen/printer_python/preamble.py:
```diff
@@ -16,364 +15,25 @@
+    @classmethod
+    def _check_retval_(cls, result):
+        value = result.value
+        return value if value is None else _bytes_wrapper(value)
```
I guess this isn't ideal since the caller looses access to the original `c_char_p`.
Maybe it would be better to return a wrapper class around the raw `c_char_p` instead, with `__getattr__()` to the value, if that works?

mara004 replied:
I tried this locally, but then methods like `upper()` return new bytes objects, which is kind of confusing.
I pushed it for the record, though. Did the original string classes behave like this, too? It looks like it...

mara004 committed:
Grant the caller access to the raw c_char_p object

mara004 committed:
apply black again

mara004 commented:
@nilason @Alan-R What do you think?

mara004 commented:
For the record, @olsonse had also written a related draft: https://github.com/ctypesgen/ctypesgen/issues/77#issuecomment-570145399
I wrote this PR independently, but perhaps said post contains relevant info to evaluate/improve the implementation.

mara004 commented:
@nilason This PR is open since approx. a month now. Any comments?

nilason commented:
> @nilason This PR is open since approx. a month now. Any comments?

I've been travelling and had other obligations, so I didn't have the time to look into this yet.

mara004 commented:
Actually, I think I should change this to outsource string-specific code and include it selectively.
Then we can let the caller choose between old/new/no string support.
It will result in some more code fragmentation but it would achieve most backwards compatibility and should therefore increase acceptability of the change.[^1]

@nilason You may wait with reviewing until I have committed this change.
I guess I will yet submit a new PR rather than editing this one.

[^1]: Though it is somewhat against my personal taste, if this were my own project I'd rather prioritize cleaner code and take the breaking change. On short term the conservative approach seems more agreeable in way, but we also need to think about future maintainability...

mara004 commented:
> Actually, I think I should change this to outsource string-specific code and include it selectively.

Small update: I have stalled this work for now, since I have a lot else to do ATM, and given the fork, spending time on backporting to mainline ctypesgen provides no immediate benefit for pypdfium2.

mara004 committed:
Fix to always go through c_char_p.from_param()

As the String class inherits from c_char_p, from_param should always go through that function and return a resolved value.

Returning bytes seems to have worked for this use case due to ctypes autoconversion of bytes, but it was not formally correct.

mara004 committed:
add comment on decode() wrapper being unnecessary in principle

mara004 committed:
satisfy flake

mara004 commented:
See also https://github.com/pypdfium2-team/ctypesgen/issues/1#issuecomment-1838936073 for a further idea: configurable kind of encoding

mara004 commented:
See https://github.com/ctypesgen/ctypesgen/pull/225#issuecomment-2859257098 for a stab at an improved string helper.

~~I'm closing this PR in favor of potentially adding this to pypdfium2-ctypesgen in the future.~~

mara004 committed:
Check in updated helper for testing

mara004 committed:
Implement an `__eq__` method

mara004 commented:
OK, nevermind, I'll keep this open for now, because it's useful for testing, and closed PRs don't take in new commits.
Sorry for the confusion.

mara004 committed:
eq: check type

mara004 committed:
improve comparison

mara004 commented:
Now implemented in pypdfium2-ctypesgen in https://github.com/pypdfium2-team/ctypesgen/commit/3bcd9e46d82275da6ca9087649fe547cb7ed90a8
